ca_gazebo/src/color_sensor_plugin.cpp

```diff
@@ -0,0 +1,121 @@
+  GZ_REGISTER_SENSOR_PLUGIN(GazeboRosColor)
+
+  ////////////////////////////////////////////////////////////////////////////////
+  // Constructor
+  GazeboRosColor::GazeboRosColor():
+    nh_("color_sensor_plugin")
+{
+
+}
+  ////////////////////////////////////////////////////////////////////////////////
+  // Destructor
+  GazeboRosColor::~GazeboRosColor()
+  {
+    ROS_DEBUG_STREAM_NAMED("camera","Unloaded");
+  }
+
+  void GazeboRosColor::Load(sensors::SensorPtr _parent, sdf::ElementPtr _sdf)
+  {
+    // Make sure the ROS node for Gazebo has already been initialized
+    if (!ros::isInitialized())
+    {
+      ROS_FATAL_STREAM("A ROS node for Gazebo has not been initialized, unable to load plugin. "
+        << "Load the Gazebo system plugin 'libgazebo_ros_api_plugin.so' in the gazebo_ros package)");
+      return;
+    }
+
+    CameraPlugin::Load(_parent, _sdf);
+    // copying from CameraPlugin into GazeboRosCameraUtils
+    this->parentSensor_ = this->parentSensor;
+    this->width_ = this->width;
+    this->height_ = this->height;
+    this->depth_ = this->depth;
+    this->format_ = this->format;
+    this->camera_ = this->camera;
+    this->rvalue_max_ = _sdf->Get<double>("rvalue_max");
+    this->rvalue_min_ = _sdf->Get<double>("rvalue_min");
+    this->gvalue_max_ = _sdf->Get<double>("gvalue_max");
+    this->gvalue_min_ = _sdf->Get<double>("gvalue_min");
+    this->bvalue_max_ = _sdf->Get<double>("bvalue_max");
+    this->bvalue_min_ = _sdf->Get<double>("bvalue_min");
+    this->publisher_name_ = _sdf->Get<std::string>("PublisherName");
+    this->sensor_publisher_ = nh_.advertise<std_msgs::Bool>(this->publisher_name_, 1);
+    this->color_percentage_=_sdf->Get<double>("ColorPercentage");
+    
+//    this->last_update_time_ = this->world_->SimTime();
+
+    GazeboRosCameraUtils::Load(_parent, _sdf);
+  }
+
+  ////////////////////////////////////////////////////////////////////////////////
+  // Update the controller
+  void GazeboRosColor::OnNewFrame(const unsigned char *_image,
+    unsigned int _width, unsigned int _height, unsigned int _depth,
+    const std::string &_format)
+  {
+
+    int R,G,B;
+
+    double desired_color_in_pixel_=0, pixel_count_=0;
+
+    this->sensor_update_time_ = this->parentSensor_->LastUpdateTime();
+
+    std_msgs::BoolPtr publ(new std_msgs::Bool);
+    
+    /*if (!this->parentSensor->IsActive())
+    {
+      if ((*this->image_connect_count_) > 0)
+      // do this first so there's chance for sensor to run once after activated
+        this->parentSensor->SetActive(true);
+    }
+    else
+    {
+      if ((*this->image_connect_count_) > 0)
+      {*/
+    const common::Time cur_time = this->world_->SimTime();
+    if (cur_time - this->last_update_time_ >= this->update_period_)
+    {
+      this->PutCameraData(_image);
+      this->PublishCameraInfo();
+      this->last_update_time_ = cur_time;
+
+      bool output_;
+
+      for (int i=0; i<_width*_height*3 ;i=i+3)
```

Comment:
Spaces between operations (e.g. `i = 0; i < _width (...)`), and also add this way: `i += 3`

Reply:
Address this comment and others too.